web/apps/apps_test.go

```diff
@@ -408,6 +411,106 @@ func TestThemeWithContext(t *testing.T) {
 	assert.NotContains(t, string(body), "/assets/styles/theme.css")
 }
 
+func TestUninstallAppWithLinkedClient(t *testing.T) {
+	// Install drive app
+	installer, err := apps.NewInstaller(testInstance, testInstance.AppsCopier(apps.Webapp),
+		&apps.InstallerOptions{
+			Operation:  apps.Install,
+			Type:       apps.Webapp,
+			Slug:       "drive",
+			SourceURL:  "registry://drive",
+			Registries: testInstance.Registries(),
+		},
+	)
+	assert.NoError(t, err)
+	_, err = installer.RunSync()
+	assert.NoError(t, err)
+
+	// Link an OAuthClient to drive
+	oauthClient := &oauth.Client{
+		ClientName:   "test-linked",
+		RedirectURIs: []string{"https://foobar"},
+		SoftwareID:   "registry://drive",
+	}
+
+	oauthClient.Create(testInstance)
+	scope := "io.cozy.apps:ALL"
+	token, err := testInstance.MakeJWT("cli", "drive", scope, "", time.Now())
+	assert.NoError(t, err)
+
+	// Trying to remove this app
+	req, _ := http.NewRequest("DELETE", ts.URL+"/apps/drive", nil)
+	req.Host = testInstance.Domain
+
+	req.Header.Add("Authorization", "Bearer "+token)
+	res, err := client.Do(req)
+	assert.NoError(t, err)
+	assert.Equal(t, 400, res.StatusCode)
+	body, err := ioutil.ReadAll(res.Body)
+	assert.NoError(t, err)
+	assert.Contains(t, string(body), "linked OAuth client exists")
+
+	// Removing the client and uninstall the app to cleanly finish the test
+	uninstaller, err := apps.NewInstaller(testInstance, testInstance.AppsCopier(apps.Webapp),
+		&apps.InstallerOptions{
+			Operation:  apps.Delete,
+			Type:       apps.Webapp,
+			Slug:       "drive",
+			SourceURL:  "registry://drive",
+			Registries: testInstance.Registries(),
+		},
+	)
+	assert.NoError(t, err)
+	uninstaller.RunSync()
+
+}
+
+func TestUninstallAppWithoutLinkedClient(t *testing.T) {
+	instance.Destroy("testinstancewithoutlinked")
+	inst, err := instance.Create(&instance.Options{
+		Domain: "testinstancewithoutlinked",
+		Locale: "en",
+	})
```

Comment:
In general, we try to not create/destroy instance in a test, but to do that in the main. At least, we should try to destroy the instance at the end of the test if it is successful (and it's better if we can also delete the instance when the test fails) to avoid polluting the CouchDB for the developers.